index.bs

```diff
@@ -4109,165 +4310,1238 @@ To <dfn>get the response data</dfn> given |response|:
+
+    1. Let |scheme| be [=ASCII lowercase=] with |pattern url|.
+
+    1. Append "<code>:</code>" to |pattern url|.
+
+    1. If |scheme| [=is special=], append "<code>//</code>" to |pattern url|.
+
+    1. If |pattern| [=map/contains=] "<code>hostname</code>":
+
+      1. If |pattern|["<code>hostname</code>"] is the empty string, return [=error=]
+         with [=error code=] [=invalid argument=].
+
+      1. If |scheme| is "<code>file</code>" return [=error=] with [=error code=]
+         [=invalid argument=].
+
+      1. Let |hostname| be the result of [=trying=] to [=unescape
+         URL Pattern=] with |pattern|["<code>hostname</code>"].
+
+      1. Let |inside brackets| be false.
+
+      1. For each |codepoint| in |hostname|:
+
+        1. If «U+002F (/), U+003F (?), U+0023 (#)» [=set/contains=]
+           |codepoint|:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+        1. If |inside brackets| is false and |codepoint| is U+003A (:):
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+        1. If |codepoint| is U+005B ([), set |inside brackets| to true.
+
+        1. If |codepoint| is U+005D (]), set |inside brackets| to false.
+
+      1. Append |hostname| to |pattern url|.
+
+    1. Otherwise:
+
+      1. If |scheme| is not "<code>file</code>", append "<code>placeholder</code>" to
+         |pattern url|.
+
+      1. Set |has hostname| to false.
+
+    1. If |pattern| [=map/contains=] "<code>port</code>":
+
+      1. If |pattern|["<code>port</code>"] is the empty string, return [=error=]
+         with [=error code=] [=invalid argument=].
+
+      1. Let |port| be the result of [=trying=] to [=unescape
+         URL Pattern=] with |pattern|["<code>port</code>"].
+
+      1. Append "<code>:</code>" to |pattern url|.
+
+      1. For each |codepoint| in |port|:
+
+        1. If |codepoint| is not an [=ASCII digit=]:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |port| to |pattern url|.
+
+    1. If |pattern| [=map/contains=] "<code>pathname</code>":
+
+      1. Let |pathname| be the result of [=trying=] to [=unescape URL Pattern=] with
+         |pattern|["<code>pathname</code>"].
+
+      1. If |pathname| does not [=string/starts with|start with=] U+002F (/), then
+         append "<code>/</code>" to pattern.
+
+      1. For each |codepoint| in |pathname|:
+
+        1. If «U+003F (?), U+0023 (#)» [=set/contains=]
+           |codepoint|:
+
+          1. Return [=error=] with [=error code=] [=invalid argument=].
+
+      1. Append |pathname| to |pattern url|.
+
+    1. If |pattern| [=map/contains=] "<code>search</code>":
+
+      1. Let |search| be the result of [=trying=] to [=unescape URL pattern=] with
+         |pattern|["<code>search</code>"].
+
+      1. If |search| does not [=string/starts with|start with=] U+003F (?), then
+         append "<code>?</code>" to |pattern url|.
```

Comment:
(just thinking out loud...) For the edge case where users want to use query parameters named `?`, they might pass `?=1`, then we won't add the `?`. Meaning the  the user won't have the expected search string in the end...

But it's quite hard to guess what the user really meant in that case, and using other default behaviors (eg always adding `?`, or never adding `?`) also can lead to confusing situations where it will be hard to understand the issue both for us and the users. In summary I think your current approach is the best compromise.